Remotion/ObjectBinding/Web.Development.WebTesting/ControlObjects/BocAutoCompleteReferenceValueControlObject.cs

```diff
@@ -392,6 +410,24 @@ private IReadOnlyList<SearchServiceResultItem> ParseSearchServiceResponse ([NotN
       }
     }
 
+    private static IReadOnlyList<SearchServiceResultItem> ParseSearchServiceResultVariants (IReadOnlyDictionary<string, object> data) =>
+        (data["Type"] as string) switch
+        {
+            "ValueList" => ParseSearchServiceResultItems(((IReadOnlyCollection<object?>)data["Values"]).Cast<IReadOnlyDictionary<string, object>>()),
+            var unknown => throw new InvalidOperationException($"The result variant '{unknown}' is not supported."),
+        };
+
+    private static IReadOnlyList<SearchServiceResultItem> ParseSearchServiceResultItems (IEnumerable<IReadOnlyDictionary<string, object>?> data) =>
+        data.Select(ParseSearchServiceResultItem).OfType<SearchServiceResultItem>().ToArray();
```

Comment:
I don't think I want to change this because the current code is much shorter than making the nullcheck + typing to a not-nullable type. Even so, it does feel strange to use a type check as a null filter. My reasoning is that as a reader I don't know from reading this particular line that we're not missing type errors via the soft-cast. E.g. if the `Parse...Item` method is re-typed to retunr a base type and then introduces a different type next to the `SearchServiceResultItem`, we'd not find this until you notice that items are missing. 

So, why not change this? basically, the naming makes things reasonable noticable and it's unlikley that this particular change would actually pop up in the future. So, right now, right here, that's probably a good thing. In the future, I think a dedidacted filter-to-not-null and convert-to-not-null extension method would probably be a good idea. thoughts?

Reply:
I guess just using `.Where(i => i != null).Cast<>()` would be enough if more clarity is wanted. Still, `OfType` is a known method with known behavior imo. I'm not sure more clarity is needed here.

Reply:
Strictly speaking, `TypeOf<>` isn't a null-filter from the APIs primairy usecase. Instead, filtering null is a sideeffct next to the type filter. In particular, it tells us that we will ignore items with a not-matching type which in turn can mask bugs intrudced by changes to the type hierarchy. 

For future stuff: would you prefer `.Where(i => i != null).Cast<>()` or `.WhereNotNull()`?

Reply:
The same applies for the `is` operator then, which means `s is string` should be expanded to `s != null && s is string` right?

> For future stuff: would you prefer .Where(i => i != null).Cast<>() or .WhereNotNull()?

I would prefer Linq methods only in this case.

Reply:
> The same applies for the `is` operator then, which means `s is string` should be expanded to `s != null && s is string` right?

Nope. There's two sceanarios:
1) ```object s;
if (s is string)```
and
1) ```object s;
if (s as string x) { return x.Length; }```

is sufficient because we need a typetest and it's okay to unify nul land the type test, even more so than before given nullable reference types.

2) ```string? s;
if (string != null) { return s.Length; }```
here, we know the type and just make this not-null.

The TypeOf scenario is only "needed" (or looking good) because there is no short way of making a null-filter and and re-typing the item from nullable to not-nullable. So, I just don't like to use a soft-cast as a null-test only instead of also a type-filter.

Reply:
Still, how is `if (s is string x)` different here, it also does not show the null check explicitly, thus should be expanded to `if (s is string x && x != null)`. That's just a matter of principle - should language features & standard API's with all their implication be required to be known or not.

Reply:
The difference is that in `if (s is string x)` the primary intention is not to do a null check, that's just an implied sideeffect. If I want to do a nullcheck here, I write a nullcheck and if I want a soft cast, I do a soft cast and it's okay and readable that the softcast also does a null check. However, if I don't need the cast but just the nullcheck, I would not write a softcast because it changes the implied semantics of the code. 

TL;DR: if I want a cast (and optional null check), I should write a cast. If I want a nullcheck, I should write a nullcheck.